### ISSUE TITLE
Cannot get notifications to work

### ISSUE THREAD

**Issue by b0bbybacc** [Icinga/icingaweb2-module-director#969]
## Expected Behavior

I should receive a notification via email when host and/or services change states 24x7x365.

## Current Behavior

I receive no notification.

## Resolution Attempts

As I am new/testing icinga2, icingaweb2, and director this stand-up was very vanilla. We currently use Nagios core and are having scale issues. 

I hit several bumps along the way and found several resources that appeared to be resolutions people found regarding setting up notifications; specifically this:

https://www.unixe.de/icinga-2-icinga-director-and-notifications/

- I have confirmed the scripts are in the proper location with the proper permissions
- I have confirmed that I am able to use sendmail (or mail) to successfully send messages from cli of host
- I am no longer receiving any warnings or errors in my deployment startup logs:

```shell
information/cli: Icinga application loader (version: v2.6.3)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: icinga.domain.tld
information/ConfigItem: Instantiated 1 ApiUser.
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 2 Zones.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 1 Endpoint.
information/ConfigItem: Instantiated 3 Notifications.
information/ConfigItem: Instantiated 2 NotificationCommands.
information/ConfigItem: Instantiated 201 CheckCommands.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 1 Host.
information/ConfigItem: Instantiated 1 User.
information/ConfigItem: Instantiated 1 UserGroup.
information/ConfigItem: Instantiated 1 TimePeriod.
information/ConfigItem: Instantiated 2 Services.
information/ConfigItem: Instantiated 1 ExternalCommandListener.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ConfigItem: Instantiated 1 IdoMysqlConnection.
information/ConfigItem: Instantiated 1 PerfdataWriter.
information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```
As I am new to this I did not test notifications specifically defined in icinga2 config files as the point is to test and see if doing everything 100% through icingaweb2 with director will work for us. 

I really hope that this is some monday detail because I am stuck.

See below for current config.

#### zones.d/director-global/timeperiods.conf
```shell
object TimePeriod "24x7x365" {
    import "legacy-timeperiod"
    import "24x7x365_template"

    display_name = "24x7x365"
    ranges = {
        "friday"	= "00:00-24:00"
        "monday"	= "00:00-24:00"
        "saturday"	= "00:00-24:00"
        "sunday"	= "00:00-24:00"
        "thursday"	= "00:00-24:00"
        "tuesday"	= "00:00-24:00"
        "wednesday"	= "00:00-24:00"
    }
}
```
#### zones.d/director-global/timeperiod_templates.conf
```shell
template TimePeriod "24x7x365_template" {
    import "legacy-timeperiod"
    ranges = {
        "friday"	= "00:00-24:00"
        "monday"	= "00:00-24:00"
        "saturday"	= "00:00-24:00"
        "sunday"	= "00:00-24:00"
        "thursday"	= "00:00-24:00"
        "tuesday"	= "00:00-24:00"
        "wednesday"	= "00:00-24:00"
    }
}
```

#### zones.d/icinga.domain.tld/notification_templates.conf
```shell
template Notification "24x7x365_email_service" {
    command = "Service Alarm By Email"
    interval = 0s
    period = "24x7x365"
    vars.icingaweb2url = "https://icinga.domain.tld/"
    vars.notification_from = "Icinga Monitor <noreply@icinga.domain.tld>"
    vars.notification_logtosyslog = "No"
}
```
#### zones.d/icinga.domain.tld/notification_templates.conf
```shell
template Notification "24x7x365_email_host" {
    command = "Host Alarm By Email"
    interval = 0s
    period = "24x7x365"
    vars.icingaweb2url = "https://icinga.domain.tld"
    vars.notification_from = "Icinga Monitor <noreply@icinga.domain.tld>"
    vars.notification_logtosyslog = "No"
}
```
#### zones.d/icinga.domain.tld/notification_apply.conf
```shell
apply Notification "Data-Center-Staff-24x7x365_host_notifier" to Host {
    import "24x7x365_email_host"

    interval = 0s
    assign where host.enable_notifications != "false"
    states = [ Down, Up ]
    types = [
        Acknowledgement,
        Custom,
        DowntimeEnd,
        DowntimeRemoved,
        DowntimeStart,
        FlappingEnd,
        FlappingStart,
        Problem,
        Recovery
    ]
    user_groups = [ "DataCenterStaff" ]
}
```
#### zones.d/icinga.domain.tld/notification_apply.conf
```shell
apply Notification "Data-Center-Staff-24x7x365_service_notifier" to Service {
    import "24x7x365_email_service"

    interval = 0s
    assign where service.enable_notifications != "false"
    states = [ Critical, OK, Unknown, Warning ]
    types = [
        Acknowledgement,
        Custom,
        DowntimeEnd,
        DowntimeRemoved,
        DowntimeStart,
        FlappingEnd,
        FlappingStart,
        Problem,
        Recovery
    ]
    user_groups = [ "DataCenterStaff" ]
}
```

#### zones.d/icinga.domain.tld/users.conf
```shell
object User "testuser" {
    import "Data Center Staff"

    display_name = "Test User"
    email = "testuser@domain.tld"
    groups = [ "DataCenterStaff" ]
}
```

#### zones.d/icinga.domain.tld/usergroups.conf
```shell
object UserGroup "DataCenterStaff" {
    display_name = "Data Center Staff"
}
```

#### zones.d/director-global/user_templates.conf
```shell
template User "Data Center Staff" {
    enable_notifications = true
    states = [ Critical, Down, OK, Unknown, Up, Warning ]
    types = [
        Acknowledgement,
        Custom,
        DowntimeEnd,
        DowntimeRemoved,
        DowntimeStart,
        FlappingEnd,
        FlappingStart,
        Problem,
        Recovery
    ]
    groups = [ "DataCenterStaff" ]
}
```

#### zones.d/director-global/commands.conf
```shell
object NotificationCommand "Host Alarm By Email" {
    import "plugin-notification-command"
    command = [ "/etc/icinga2/scripts/host-by-mail.sh" ]
    arguments += {
        "-4" = {
            required = true
            value = "$address$"
        }
        "-6" = "$address6$"
        "-b" = "$notification.author$"
        "-c" = "$notification.comment$"
        "-d" = {
            required = true
            value = "$icinga.long_date_time$"
        }
        "-f" = "$notification_from$"
        "-i" = "$icingaweb2url$"
        "-l" = {
            required = true
            value = "$host.name$"
        }
        "-n" = {
            required = true
            value = "$host.display_name$"
        }
        "-o" = {
            required = true
            value = "$host.output$"
        }
        "-r" = {
            required = true
            value = "$user.email$"
        }
        "-s" = {
            required = true
            value = "$host.state$"
        }
        "-t" = {
            required = true
            value = "$notification.type$"
        }
        "-v" = "$notification_logtosyslog$"
    }
}
```

#### zones.d/director-global/commands.conf
```shell
object NotificationCommand "Service Alarm By Email" {
    import "plugin-notification-command"
    command = [ "/etc/icinga2/scripts/service-by-mail.sh" ]
    arguments += {
        "-4" = {
            required = true
            value = "$address$"
        }
        "-6" = "$address6$"
        "-b" = "$notification.author$"
        "-c" = "$notification.comment$"
        "-d" = {
            required = true
            value = "$icinga.long_date_time$"
        }
        "-e" = {
            required = true
            value = "$service.name$"
        }
        "-f" = "$notification_from$"
        "-i" = "$icingaweb2url$"
        "-l" = {
            required = true
            value = "$host.name$"
        }
        "-n" = {
            required = true
            value = "$host.display_name$"
        }
        "-o" = {
            required = true
            value = "$service.output$"
        }
        "-r" = {
            required = true
            value = "$user.email$"
        }
        "-s" = {
            required = true
            value = "$service.state$"
        }
        "-t" = "$notification.type$"
        "-u" = {
            required = true
            value = "$service.display_name$"
        }
        "-v" = "$notification_logtosyslog$"
    }
}
```

## Host environment

* Director version (System - About):

> Name	director
State	enabled disable
Version	1.3.1
Git commit	4d240976a3e1ffe1eb9555fa23067da4aed25a58
Description	Director - Config tool for Icinga 2
Icinga Director is a configuration tool that has been designed to make Icinga 2 configuration easy and understandable.
Dependencies	monitoring
Permissions	director/api: Allow to access the director API
director/audit: Allow to access the full audit log
director/showconfig: Allow to show configuration (could contain sensitive information)
director/deploy: Allow to deploy configuration
director/hosts: Allow to configure hosts
director/users: Allow to configure users
director/notifications: Allow to configure notifications
director/inspect: Allow to inspect objects through the Icinga 2 API (could contain sensitive information)
director/*: Allow unrestricted access to Icinga Director

* Icinga Web 2 version and modules (System - About):

>Version
2.4.1
Git commit
d74124bd12f53d13b5a6ea5408e9783cc5190d1e
Git commit date
2017-01-18
Copyright
© 2013-2017 The Icinga Project

* Icinga 2 version (`icinga2 --version`):

>icinga2 - The Icinga 2 network monitoring daemon (version: v2.6.3)
Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

* Operating System and version:

>System information:
  Platform: CentOS Linux
  Platform version: 7 (Core)
  Kernel: Linux
  Kernel version: 3.10.0-514.21.1.el7.x86_64
  Architecture: x86_64
Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

* Webserver, PHP versions:

>PHP 5.4.16 (cli) (built: Nov  6 2016 00:29:02)
Copyright (c) 1997-2013 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2013 Zend Technologies

>Server version: Apache/2.4.6 (CentOS)
Server built:   Apr 12 2017 21:03:28

**Comment by Thomas-Gelf** [comment]
Hi @b0bbybacc,

is the notification feature (`icinga2 feature list`)enabled on your Icinga 2 host? Anything notification-related in your `icinga2.log`? Did a related host reach a "hard state" and did a related state change occur? Does Icinga Web 2 mention sent notifications on the host detail page? Or when you go to it's history? Could you eventually try to set a notification interval to have recurring notifications (just to have a better chance to catch something in your logs)?

Your startup log shows one host, two services and three notifications - this means that your apply rules should match.

Best,
Thomas

**Comment by b0bbybacc** [comment]
Thomas,

Thanks for this. The second I read

>Is the notification feature (icinga2 feature list)enabled on your Icinga 2 host?

I **knew** that was it as I disabled it sometime yesterday trying to work this issue out and did not enable it after today's discovery of the article from my initial comment. 

Absolutely a monday detail and thank you ever so much for saving me from more random head-bashing!

**Comment by Thomas-Gelf** [comment]
You're welcome :-)

**Comment by blarg3891** [comment]
> Thomas,
> 
> Thanks for this. The second I read
> 
> > Is the notification feature (icinga2 feature list)enabled on your Icinga 2 host?
> 
> I **knew** that was it as I disabled it sometime yesterday trying to work this issue out and did not enable it after today's discovery of the article from my initial comment.
> 
> Absolutely a monday detail and thank you ever so much for saving me from more random head-bashing!

Hi,
I just setup the Icinga2 docker which includes director and so forth in there.  I have been trying to figure out how to get notifications to work.  Generally my configs look the same as yours after I followed the link https://www.unixe.de/icinga-2-icinga-director-and-notifications/.

I also have mail through ssmtp working as I have tested in the commandline.  I cannot for the life of me figure out how to make notifications show up for my new services that I created in director.

I was hoping maybe you had a help file you created for yourself when you set this up...

Any help at all would be much appreciated.  Been banging my head against the wall.
